src/textual/widgets/_progress_bar.py

```diff
@@ -0,0 +1,432 @@
+                background_style=Style.from_color(bar_style.bgcolor),
+            )
+
+    def render_indeterminate(self) -> RenderResult:
+        """Render a frame of the indeterminate progress bar animation."""
+        width = self.size.width
+        highlighted_bar_width = 0.25 * width
+        # Width used to enable the visual effect of the bar going into the corners.
+        total_imaginary_width = width + highlighted_bar_width
+
+        speed = 30  # Cells per second.
+        # Compute the position of the bar.
+        start = (speed * monotonic()) % (2 * total_imaginary_width)
+        if start > total_imaginary_width:
+            # If the bar is to the right of its width, wrap it back from right to left.
+            start = 2 * total_imaginary_width - start  # = (tiw - (start - tiw))
+        start -= highlighted_bar_width
+        end = start + highlighted_bar_width
+
+        bar_style = self.get_component_rich_style("bar--indeterminate")
+        return RenderableBar(
+            highlight_range=(max(0, start), min(end, width)),
+            highlight_style=Style.from_color(bar_style.color),
+            background_style=Style.from_color(bar_style.bgcolor),
+        )
+
+
+class PercentageStatus(Label):
+    """A label to display the percentage status of the progress bar."""
+
+    _completion_percentage: reactive[float | None] = reactive[Optional[float]](None)
+    """The percentage of progress that has been completed or `None` if indeterminate."""
+
+    DEFAULT_CSS = """
+    PercentageStatus {
+        width: 5;
+        content-align-horizontal: right;
+    }
+    """
+
+    def watch__completion_percentage(self, percentage: float | None) -> None:
+        if percentage is None:
+            self.renderable = "--%"
+        else:
+            self.renderable = f"{int(100 * percentage)}%"
+
+
+class ETAStatus(Label):
+    """A label to display the estimated time until completion of the progress bar."""
+
+    _completion_percentage: reactive[float | None] = reactive[Optional[float]](None)
+    """The percentage of progress that has been completed or `None` if indeterminate."""
+
+    DEFAULT_CSS = """
+    ETAStatus {
+        width: 9;
+        content-align-horizontal: right;
+    }
+    """
+
+    _refresh_timer: Timer | None = None
+    """Timer to update ETA status even when progress stalls."""
+    _start_time: float | None = None
+    """The time at which the progress started being tracked."""
+
+    def watch__completion_percentage(self, percentage: float | None) -> None:
+        if percentage is None:
+            self.renderable = "--:--:--"
+        else:
+            # Keep track of the starting time.
+            if self._start_time is None:
+                self._start_time = monotonic()
+            # If we are not notified of progress for some time, periodically refresh
+            # the countdown so that the ETA is updated either way.
+            if self._refresh_timer is None:
```

Comment:
It makes me nervous creating objects in watch functions. Looks fine, but the code implies you will be creating new objects here.

Suggest we move setting the interval outside of the watch function. How about create it `on_mount`? That way it is clear there will only be a single instance.